backend/fms_core/services/sample_lineage.py

```diff
@@ -25,3 +26,82 @@ def create_sample_lineage(parent_sample, child_sample, process_measurement):
             errors.append(str(e))
 
     return (sample_lineage, errors, warnings)
+
+def create_sample_lineage_graph(mid_sample):
+    errors = []
+
+    if not mid_sample:
+        errors.append(f"Sample is required for sample lineage graph creation")
+
+    nodes = [mid_sample]
+    edges = []
+
+    if not errors:
+        def add_lineage(process_measurements):
+            queryset = process_measurements.select_related("process").prefetch_related("lineage")
+            queryset = queryset.annotate(child_sample=F("lineage__child"))
+            queryset = queryset.annotate(child_sample_name=F("lineage__child__name"))
+            queryset = queryset.annotate(source_sample_name=F("source_sample__name"))
+            return queryset
+
+        def query_set_iterator(f, l):
+            for x in l:
+                try:
+                    yield f(x)
+                except ObjectDoesNotExist:
+                    yield None
+
+        # fetch children
```

Comment:
I am not sure you need to start from the middle and basically repeat the operation up and down. All the samples in a lineage share the same biosample.id. So if you want all samples you can get them with a single query by testing this. To get the root sample(s), you can eliminate the ones that are child_sample in the lineage table. I haven't tracked your algorithm closely but it should be equivalent to putting on the stack all root samples in your fetch children section. This would simplify the graph traversal algorithm i think. You may even want to dodge the iterative traversal by listing all nodes using the biosample.id and listing all edges using the fact that all edges will be the ones that have a samplelineage.parent_sample and samplelineage.child sample in the node list. Up to you...

Reply:
The intention of my algorithm is that I want to explore the entire children of the middle node, but I'm not interested in all of the children of parents. My concern is that I might fetch too many samples.